sycl/include/sycl/reduction.hpp

```diff
@@ -179,7 +182,8 @@ template <typename ReducerT> class ReducerAccess {
   template <typename ReducerRelayT = ReducerT>
   enable_if_t<
       !IsKnownIdentityOp<typename ReducerRelayT::value_type,
-                         typename ReducerRelayT::binary_operation>::value,
+                         typename ReducerRelayT::binary_operation>::value &&
+          ReducerTraits<ReducerRelayT>::has_identity,
```

Comment:
Can we get one step further and just delegate all of the lines 174-201 to the `IdentityContainer`? We can provide the delegates unconditionally here expecting `static_assert` inside the `IdentityContainer` methods if something is unexpected/unsupported.

Reply:
I've made a general simplification of these. In this I've added the static asserts early in the functions to make any errors somewhat reasonable. I don't think it's worth moving any of the remaining logic to `ReductionIdentityContainer`.